From f23d211546afcd145a75a2de4f11d73e029156b7 Mon Sep 17 00:00:00 2001
From: Kirill Kapranov <kirill.kapranov@compulab.co.il>
Date: Wed, 17 Jun 2020 17:58:35 +0300
Subject: [PATCH 41/49] Fix memory size initialisation

The u-boot sets correct memory bank sizes.
Note: OP-TEE support is disabled.

Signed-off-by: Kirill Kapranov <kirill.kapranov@compulab.co.il>
---
 arch/arm/mach-imx/imx8m/soc.c                  |  9 ++++-
 board/compulab/ucm-imx8m-mini/ddr/ddr.c        | 11 +-----
 board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c | 51 +++++++++++++++-----------
 board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.h |  7 ++++
 configs/ucm-imx8m-mini_defconfig               |  2 +-
 include/configs/ucm-imx8m-mini.h               |  7 +---
 6 files changed, 49 insertions(+), 38 deletions(-)

diff --git a/arch/arm/mach-imx/imx8m/soc.c b/arch/arm/mach-imx/imx8m/soc.c
index 2b2f791348..68437c297f 100644
--- a/arch/arm/mach-imx/imx8m/soc.c
+++ b/arch/arm/mach-imx/imx8m/soc.c
@@ -169,7 +169,7 @@ struct mm_region *mem_map = imx8m_mem_map;
 void enable_caches(void)
 {
 	/* If OPTEE runs, remove OPTEE memory from MMU table to avoid speculative prefetch */
-	if (rom_pointer[1]) {
+	if (1 /*rom_pointer[1]*/) {
 
 		/* TEE are loaded, So the ddr bank structures
 		* have been modified update mmu table accordingly
@@ -179,7 +179,12 @@ void enable_caches(void)
 		* imx8m_mem_map for DRAM1
 		*/
 		int entry = 5;
-		u64 attrs = imx8m_mem_map[entry].attrs;
+		u64 attrs = PTE_BLOCK_MEMTYPE(MT_NORMAL) |
+#ifdef CONFIG_IMX_TRUSTY_OS
+				PTE_BLOCK_INNER_SHARE;
+#else
+				PTE_BLOCK_OUTER_SHARE;
+#endif
 		while (i < CONFIG_NR_DRAM_BANKS && entry < 8) {
 			if (gd->bd->bi_dram[i].start == 0)
 				break;
diff --git a/board/compulab/ucm-imx8m-mini/ddr/ddr.c b/board/compulab/ucm-imx8m-mini/ddr/ddr.c
index 0828917cc3..3cf9f016e0 100644
--- a/board/compulab/ucm-imx8m-mini/ddr/ddr.c
+++ b/board/compulab/ucm-imx8m-mini/ddr/ddr.c
@@ -13,6 +13,7 @@
 #include <asm/arch/clock.h>
 #include <asm/mach-imx/gpio.h>
 #include "ddr.h"
+#include "../ucm-imx8m-mini.h"
 
 /* Forward declarations */
 u32 cl_eeprom_get_ddrinfo(void);
@@ -59,13 +60,6 @@ struct lpddr4_desc {
 	char *desc[4];
 };
 
-struct lpddr4_tcm_desc {
-	unsigned int size;
-	unsigned int sign;
-	unsigned int index;
-	unsigned int count;
-};
-
 #define DEFAULT (('D' << 24) + ('E' << 16 ) + ( 'F' << 8 ) + 'A')
 static const struct lpddr4_desc lpddr4_array[] = {
 	{ .name = "Micron", .id = 0xff020008, .size = 2048, .count = 1, .timing = &ucm_dram_timing_ff020008},
@@ -110,7 +104,6 @@ static void spl_tcm_fini(struct lpddr4_tcm_desc *lpddr4_tcm_desc) {
     lpddr4_tcm_desc->index = 0;
 }
 
-#define SPL_TCM_DATA 0x7e0000
 #define SPL_TCM_INIT spl_tcm_init(lpddr4_tcm_desc)
 #define SPL_TCM_FINI spl_tcm_fini(lpddr4_tcm_desc)
 
@@ -121,7 +114,7 @@ void spl_dram_init(void)
 	unsigned int ddr_found = 0;
 	int i = 0;
 
-	struct lpddr4_tcm_desc *lpddr4_tcm_desc = (struct lpddr4_tcm_desc *) SPL_TCM_DATA;
+	struct lpddr4_tcm_desc *lpddr4_tcm_desc = (struct lpddr4_tcm_desc *) TCM_DATA_CFG;
 
 	if (lpddr4_tcm_desc->sign != DEFAULT) {
 		/* get ddr type from the eeprom if not in tcm scan mode */
diff --git a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
index fa06db9ec6..08b267fc01 100644
--- a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
+++ b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
@@ -153,25 +153,21 @@ int board_postclk_init(void)
 #endif
 int board_phys_sdram_size(phys_size_t *size)
 {
-    unsigned long value = readl(TCM_DATA_CFG);
-    *size = 0x80000000;
-
-    switch (value) {
-    case 4096:
-        /*value = 3084;*/
-    case 3084:
-    case 2048:
-    case 1536:
-    case 1024:
-    case 768:
-    case 512:
-    case 256:
-        *size = ( value << 20 );
-        break;
-    default:
-        break;
-    };
-    return 0;
+	struct lpddr4_tcm_desc *lpddr4_tcm_desc = (struct lpddr4_tcm_desc *) TCM_DATA_CFG;
+
+	switch (lpddr4_tcm_desc->size) {
+	case 4096:
+	case 2048:
+	case 1024:
+		*size = (1L << 20) * lpddr4_tcm_desc->size;
+		break;
+	default:
+		printf("%s: DRAM size %uM is not supported \n", __func__,
+				lpddr4_tcm_desc->size);
+		while ( 1 ) {};
+		break;
+	};
+	return 0;
 }
 	/* Get the top of usable RAM */
 ulong board_get_usable_ram_top(ulong total_size)
@@ -188,6 +184,7 @@ int dram_init(void)
 	phys_size_t sdram_size;
 	int ret;
 	ret = board_phys_sdram_size(&sdram_size);
+	printf("%s: size %llu\n", __func__, sdram_size);
 	if (ret)
 		return ret;
 
@@ -219,21 +216,31 @@ int dram_init_banksize(void)
 	}
 
 	gd->bd->bi_dram[bank].start = PHYS_SDRAM;
+	gd->bd->bi_dram[bank].size = bank_1_size;
+
 	if (rom_pointer[1]) {
 		phys_addr_t optee_start = (phys_addr_t)rom_pointer[0];
 		phys_size_t optee_size = (size_t)rom_pointer[1];
 
-		gd->bd->bi_dram[bank].size = optee_start -gd->bd->bi_dram[bank].start;
+		
 		if ((optee_start + optee_size) < (PHYS_SDRAM + bank_1_size)) {
+
+			gd->bd->bi_dram[bank].size = optee_start -
+			       	gd->bd->bi_dram[bank].start;
+
 			if ( ++bank >= CONFIG_NR_DRAM_BANKS) {
 				puts("CONFIG_NR_DRAM_BANKS is not enough\n");
 				return -1;
 			}
 
-			gd->bd->bi_dram[bank].start = optee_start + optee_size;
+			gd->bd->bi_dram[bank].start = optee_start;
 			gd->bd->bi_dram[bank].size = PHYS_SDRAM +
 				bank_1_size - gd->bd->bi_dram[bank].start;
 		}
+		else {
+			printf("%s: There is no room to allocate TEE 0x%16.16llx @ 0x%16.16llx\n", __func__, optee_size, optee_start);
+		}
+
 	} else {
 		gd->bd->bi_dram[bank].size = bank_1_size;
 	}
@@ -245,7 +252,9 @@ int dram_init_banksize(void)
 		}
 		gd->bd->bi_dram[bank].start = PHYS_SDRAM_2;
 		gd->bd->bi_dram[bank].size = bank_2_size;
+	printf("%s:%i %lu %llu [%i] %llu\n", __func__, __LINE__, gd->bd->bi_dram[bank].start, gd->bd->bi_dram[bank].size, bank, total_size);
 	}
+	printf("%s:%i %lu %llu [%i] %llu\n", __func__, __LINE__, gd->bd->bi_dram[bank].start, gd->bd->bi_dram[bank].size, bank, total_size);
 	return 0;
 }
 
diff --git a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.h b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.h
index be698941da..55bdadd3a6 100644
--- a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.h
+++ b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.h
@@ -9,4 +9,11 @@
 
 #define TCM_DATA_CFG 0x7e0000
 
+struct lpddr4_tcm_desc {
+	unsigned int size;
+	unsigned int sign;
+	unsigned int index;
+	unsigned int count;
+};
+
 #endif
diff --git a/configs/ucm-imx8m-mini_defconfig b/configs/ucm-imx8m-mini_defconfig
index c57d88fa59..f42597b5c7 100644
--- a/configs/ucm-imx8m-mini_defconfig
+++ b/configs/ucm-imx8m-mini_defconfig
@@ -5,7 +5,7 @@ CONFIG_SYS_TEXT_BASE=0x40200000
 CONFIG_SYS_MALLOC_F_LEN=0x2000
 CONFIG_TARGET_UCM_IMX8M_MINI=y
 CONFIG_SPL_MMC_SUPPORT=y
-CONFIG_IMX_OPTEE=y
+# CONFIG_IMX_OPTEE is not set
 CONFIG_DEFAULT_DEVICE_TREE="ucm-imx8m-mini"
 CONFIG_FIT=y
 CONFIG_SPL_LOAD_FIT=y
diff --git a/include/configs/ucm-imx8m-mini.h b/include/configs/ucm-imx8m-mini.h
index 30caaca5a1..1d1108a6b9 100644
--- a/include/configs/ucm-imx8m-mini.h
+++ b/include/configs/ucm-imx8m-mini.h
@@ -207,13 +207,10 @@
 
 #define PHYS_SDRAM              0x40000000
 #define PHYS_SDRAM_2            0x100000000
-#define PHYS_SDRAM_SIZE         0x80000000 /* 2GB default size*/
-#undef PHYS_SDRAM_2_SIZE /* Memory chip autodetection */
+#define PHYS_SDRAM_SIZE   0 /* Memory chip autodetection */
+#define PHYS_SDRAM_2_SIZE 0 /* Memory chip autodetection */
 #define CONFIG_NR_DRAM_BANKS    4
-
 #define CONFIG_SYS_SDRAM_BASE   PHYS_SDRAM
-#define PHYS_SDRAM_SIZE			0x80000000 /* 2GB DDR */
-#define PHYS_SDRAM_2_SIZE		0x40000000
 
 #define MEMTEST_DIVIDER   2
 #define MEMTEST_NUMERATOR 1
-- 
2.11.0

